### PR TITLE
Automate releases

### LINKED ISSUE #269: Automatically add go binaries to the release
Whenever we release a new version of the CLI, we should automatically compile the binary for Linux, macos and windows, 32 and 64 bits and upload them to the Github release.

### PR DESCRIPTION
**Pull request by tie**
This PR automates publishing new releases.

To publish a new release, just `git push release`. The new version is automatically inferred based on API changes.
Alternatively, it’s possible to override this behavior (e.g. if we have breaking changes in cmd/genji but not in Go API) by pushing to `release-v0.13.0` branch, or manually dispatching a workflow with version inputs.

The workflow then creates version bump commits for Genji’s submodules and tags them. In the end it creates a draft release with changelog (additionally using `CHANGELOG.md` if it exists). When GitHub release is published, another CI workflow kicks in that builds and uploads binaries as release assets.

Additionally, version bump commits are only reachable from tags so it should be safe to dispatch the workflow on main branch.

That said, this should also eliminate the need to manually manage unreleased versions in `go.mod`.

workflow

<details>
<summary>Build matrix</summary>

<div align=center>
Build matrix
</div>
</details>

Closes #269

### CODE REVIEW

**Comment by tie** [comment]
Current status: `release-assets.yaml` workflow for building and uploading binaries for new releases is working! I haven’t tested (yet) the `release.yaml` workflow that automates creating and tagging new semver releases for all Go submodules, and also creates a release draft with the list of API changes (using [x/exp/cmd/gorelease](https://golang.org/x/exp/cmd/gorelease)).

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+          go mod edit \
+            -dropreplace=github.com/genjidb/genji \
+            -dropreplace=github.com/genjidb/engine/badgerengine
+          go get \
+            github.com/genjidb/genji@${{ steps.changelog.outputs.version }} \
+            github.com/genjidb/genji/engine/badgerengine@${{ steps.changelog.outputs.version }}
+          go mod tidy
+          go mod edit \
+            -replace=github.com/genjidb/genji=../../ \
+            -replace=github.com/genjidb/engine/badgerengine=../../
```
Comment (tie) [inline .github/workflows/release.yaml:132]:
```suggestion
            -replace=github.com/genjidb/engine/badgerengine=../../engine/badgerengine
```

File: .github/workflows/release-assets.yaml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+          path: ${{ steps.go-env.outputs.modcache }}
+          key: modcache-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: modcache-go-
+
+      - name: Set up build cache
+        uses: actions/cache@v2.1.5
+        with:
+          path: ${{ steps.go-env.outputs.cache }}
+          key: cache-go-${{ matrix.os }}-${{ matrix.arch }}-${{ hashFiles('**/go.sum') }}
+          restore-keys: cache-go-${{ matrix.os }}-${{ matrix.arch }}
```
Comment (tie) [inline .github/workflows/release-assets.yaml:116]:
```suggestion
          restore-keys: cache-go-${{ matrix.os }}-${{ matrix.arch }}-
```

**Review by tie — COMMENTED** [review]
Oops, adding a few typos a review change suggestions so that I can apply them from UI.

**Comment by tie** [comment]
Testing release.yaml workflow by changing the import path and creating a new release in a separate repository.

```
find . \( -type d -name .git -prune \) -o -type f -print0 | xargs -0 sed -i 's:genjidb/genji:tie/genji-release-test:g'
go mod tidy
(cd engine/badgerengine && go mod tidy)
(cd cmd/genji && go mod tidy)
```

**Comment by tie** [comment]
Looks like gorelease silently ignores replace directive. I’ll add a rollback strategy and run gorelease for badgerengine in a separate job *after* tagging root module.

**Comment by tie** [comment]
@asdine should be ready for review.

File: .github/workflows/release-assets.yaml [hunk 3]
```diff
@@ -0,0 +1,161 @@
+name: Release Assets
+on:
+  release:
+    types:
+      - published
+
+env:
+  # Ensure that we still have a chance to remove tags event after release.
```
Comment (tie) [inline .github/workflows/release-assets.yaml:8]:
```suggestion
  # Ensure that we still have a chance to remove tags even after release.
```

**Comment by asdine** [comment]
Thanks @tie ! I will review ASAP

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+          # Ensure that default merge commit body contains [skip ci] so that we don’t trigger release workflow.
+          # See also https://github.blog/changelog/2021-02-08-github-actions-skip
+          title: 'Merge release tags for ${{ needs.prepare1.outputs.version }} [skip ci]'
+          body: |
+            This PR merges ${{ needs.prepare1.outputs.version }} release tags.
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  # Remove tags and draft release on failure.
+  #
+  rollback:
```
Comment (tie) [inline .github/workflows/release.yaml:458]: We should also set concurrency group for release workflow. See https://github.blog/changelog/2021-04-19-github-actions-limit-workflow-run-or-job-concurrency/

**Review by asdine — COMMENTED** [review]
Thanks @tie ! Sorry for the delay

I don't know much about Github Actions so I don't have much to say about the manifests and I trust your expertise. (cc @jhchabran perhaps you know more than me, if you want to have a look)
@tie could you add a README explaining how to create a release? Otherwise I'm sure I'll forgot the minute I merge the PR 😆

**Comment by gedw99** [comment]
This looks good !

would you be able to add the tinygo wasm target into the matrix ?

**Comment by jhchabran** [comment]
@tie That's a lot of good work you poured into making this work! And those things are not fun to debug either, thank you so much for taking the time to make it work. 

--- 

I experimented with the process and it works really nicely. AFAIU, with this PR the development process becomes the following: 

1. Classic PR and merge onto the `main` branch, as usual 
2. When it's time to release, `git push origin main:release` to trigger the release workflow 
3. This opens a PR from `release-vX.Y.Z` to `release` where we can review the updates to all `go.mod` (and `go.sum` files). 
5. We turn the release draft that was automatically created, which triggers the release assets workflow  
6. The build matrix runs and attaches the assets 
7. Drop the release branch. 

> Additionally, version bump commits are only reachable from tags so it should be safe to dispatch the workflow on main branch.

So that means, we'd be stuck with https://github.com/genjidb/genji/pull/401/files#diff-f94dd260d5f8f1b19caefcdfefe9ad1ab0cdfc040cd655cc82367b6950da9a0dR9 "forever" unless we rebase the `release` branch onto `main. 

➡️ I don't think that's an issue, but I just wanted to be 100% sure I understand how we should handle the `release` branch. Could you confirm or correct me here? 

--- 

Short feedback on minor things: 

➡️ Optionally, what about zipping the binaries assets from the build matrix before uploading them on GH? 

➡️ Could you squash the commits before we merge this PR? There are 54 of them and most of them are internal iterarations that do not bring any value to be kept as is in the commit history. 

PS: thanks for taking the time to write this comment, it saved me some time 😊  https://github.com/genjidb/genji/pull/401#issuecomment-841368312 !

**Comment by jhchabran** [comment]
@tie Zipping things aside which is not important, could you just confirm what is the expected flow with the release branch?

**Comment by jhchabran** [comment]
After discussing with @asdine, we have decided to not merge this pull-request. This decision has nothing to do with the quality of your contribution, which is absolutely great! 

➡️ What motivates our decision is that at this stage, Genji is still small and we aren't releasing often enough to justify adding an automated release mechanism and having to own that code. 

I'll keep a bookmark to this PR because it's a great example of what can be achieved with Github actions. I hope this come out as not too disappointing due to how much energy was invested here.

**Pull request closed without merging**